manager/dispatcher/dispatcher.go

```diff
@@ -1103,6 +1116,13 @@ func (d *Dispatcher) markNodeNotReady(id string, state api.NodeStatus_State, mes
 // Node should send new heartbeat earlier than now + TTL, otherwise it will
 // be deregistered from dispatcher and its status will be updated to NodeStatus_DOWN
 func (d *Dispatcher) Heartbeat(ctx context.Context, r *api.HeartbeatRequest) (*api.HeartbeatResponse, error) {
+	d.shutdownWait.Add(1)
+	defer d.shutdownWait.Done()
+
+	if !d.isRunning() {
```

Comment:
This is the version with no lock, so can potentially race with the shutdown, is it a problem?
If not can you add a comment saying the reason why it is fine like that?

Reply:
Its not because Stop() will wait for Heardbeat() to complete since it has already incremented the waitgroup. Added a comment.

Reply:
The wait of the WG is not the first operation of the Stop so the check for running here is racing with the beginning of the stop

Reply:
as I was saying maybe is not a problem but with this change you would have start the stop function and still process an extra heartbeat because the check isRunning check the bool with no lock

Reply:
>with this change you would have start the stop function and still process an extra heartbeat because the check isRunning check the bool with no lock

True. There is no correctness issue and the win is that we don't need to get a lock in Heartbeat(), which is the most frequent call on the dispatcher.